### ISSUE TITLE
Buffer WebTransport streams and datagrams if they arrive before the response for the CONNECT request arrives

### ISSUE THREAD

**Issue by ddragana** [mozilla/neqo#1372]

**Comment by larseggert** [comment]
CC @jesup

**Comment by mxinden** [comment]
Corresponding draft section:

> Clients can, however, send a SETTINGS frame, multiple WebTransport CONNECT requests, WebTransport data streams, and WebTransport datagrams all within a single flight. As those can arrive out of order, a WebTransport server could be put into a situation where it receives a stream or a datagram without a corresponding session. Similarly, a client may receive a server-initiated stream or a datagram before receiving the CONNECT response headers from the server.To handle this case, WebTransport endpoints SHOULD buffer streams and datagrams until they can be associated with an established session.
>
> To avoid resource exhaustion, endpoints MUST limit the number of buffered streams and datagrams. When the number of buffered streams is exceeded, a stream SHALL be closed by sending a RESET_STREAM and/or STOP_SENDING with the WT_BUFFERED_STREAM_REJECTED error code. When the number of buffered datagrams is exceeded, a datagram SHALL be dropped. It is up to an implementation to choose what stream or datagram to discard.

https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#section-4.5

**Comment by mxinden** [comment]
Note that the issue title includes both streams and datagrams that might arrive early. #2806 only fixes the issue for streams.

I still think we should close here. Datagrams need to be treated as best-effort by the sender anyways. Unless we have a good use-case, I don't think doing the same fix for datagrams is worth the complexity.

Please comment if you think otherwise or have a use-case.

**Comment by afrind** [comment]
If a QUIC DATAGRAM is in the same QUIC packet as the 200 Ok, will it be correctly delivered to the application?  It would be kind of nice if this case was handled.

**Comment by mxinden** [comment]
@afrind yes, as long as the 200 OK is before the WebTransport datagram in the QUIC packet, the client will correctly handle both.

https://github.com/mozilla/neqo/pull/2979 will add a test for it.
